clusterloader2/pkg/test/simpletestexecutor.go

```diff
@@ -27,21 +35,142 @@ func createSimpleTestExecutor() TestExecutor {
 }
 
 // ExecuteTest executes test based on provided configuration.
-func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) error {
-	return nil
+func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) []error {
+	defer cleanupResources(ctx)
+	var errList []error
+	err := ctx.GetFramework().CreateAutomanagedNamespaces(int(conf.AutomanagedNamespaces))
+	if err != nil {
+		errList = append(errList, fmt.Errorf("automanaged namespaces creation failed: %v", err))
+		return errList
+	}
+
+	for i := range conf.Steps {
+		if stepErrList := ste.ExecuteStep(ctx, &conf.Steps[i]); len(stepErrList) > 0 {
+			errList = append(errList, stepErrList...)
+			if isErrsCritical(stepErrList) {
+				return errList
+			}
+		}
+	}
+	return errList
 }
 
 // ExecuteStep executes single test step based on provided step configuration.
-func (ste *simpleTestExecutor) ExecuteStep(ctx Context, step *api.Step) error {
-	return nil
+func (ste *simpleTestExecutor) ExecuteStep(ctx Context, step *api.Step) []error {
+	var errList []error
+	if len(step.Measurements) > 0 {
+		// TODO: handle measurments
+	} else {
+		var wg wait.Group
+		var lock sync.Mutex
+		for i := range step.Phases {
+			phase := &step.Phases[i]
+			wg.Start(func() {
+				if phaseErrList := ste.ExecutePhase(ctx, phase); len(phaseErrList) > 0 {
+					lock.Lock()
+					defer lock.Unlock()
+					errList = append(errList, phaseErrList...)
+				}
+			})
+		}
+		wg.Wait()
+	}
+	return errList
 }
 
 // ExecutePhase executes single test phase based on provided phase configuration.
-func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) error {
+func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) []error {
+	var errList []error
+	nsList := createNamespacesList(phase.NamespaceRange)
+	for _, nsName := range nsList {
+		for j := len(phase.ObjectBundle) - 1; j >= 0; j-- {
+			id, err := getIdentifier(&phase.ObjectBundle[j])
+			if err != nil {
+				errList = append(errList, err)
+				continue
+			}
+			instances, exists := ctx.GetState().Get(nsName, id)
+			if !exists {
+				instances = &state.InstancesState{
+					ReplicaCount: 0,
+					Object:       phase.ObjectBundle[j],
+				}
+			}
+			currentInstancesCount := instances.ReplicaCount
+			instances.ReplicaCount = phase.ReplicasPerNamespace
+			ctx.GetState().Set(nsName, id, instances)
+			// Handling for remove objects
+			for replicaIndex := phase.ReplicasPerNamespace; replicaIndex < currentInstancesCount; replicaIndex++ {
+				if objectErrList := ste.ExecuteObject(ctx, &phase.ObjectBundle[j], nsName, replicaIndex); len(objectErrList) > 0 {
+					errList = append(errList, objectErrList...)
+					if isErrsCritical(objectErrList) {
+						return errList
+					}
+				}
```

Comment:
If indeed the case is of 'removing objects' why don't you just break from here? I think it's unnecessary to go to the update/create cases.

Reply:
Let's assume that we have cluster with 20 A and 0 B objects. We want to have 10 x [A, B]. In this scenario we want to delete 10 A and create 10 B. As you can see both types of operation are used in one step.

Reply:
Yeah, that's true. However, what I meant above is that for a given object in objectbundle, you should need to do exactly one of the operations delete/create/update iiuc (and doesn't seem to be the case right now). So could you change code as follows:

```
for object in objectbundle {
 if desiredCount > currentCount {<code for create>}
 if desiredCount < currentCount {<code for delete>}
 if desiredCount == currentCount {<code for update>}
}
```